Make bats e2e tests idempotent

**Describe the solution you'd like**

The ability to run `bats -t test/bats/test.bats` repeatedly without failing. Right now only the first run succeeds, but when run a second time it produces errors because the previous run did not successfully clean up. This would be really useful to have during iterative development.


**Anything else you would like to add:**

The first time succeeds. When run again, it produces the following errors:

```shell
→ make test-e2e ENABLE_MUTATION_TESTS=y
bats -t test/bats/test.bats
<path-to-repo>/open-policy-agent/gatekeeper/test/bats
1..19
ok 1 gatekeeper-controller-manager is running
ok 2 gatekeeper-audit is running
ok 3 namespace label webhook is serving
ok 4 constrainttemplates crd is established
ok 5 mutation crds are established
ok 6 waiting for validating webhook
ok 7 gatekeeper mutation test
ok 8 applying sync config
not ok 9 create basic resources
# (in test file test/bats/test.bats, line 83)
#   `kubectl create ns gatekeeper-excluded-namespace' failed
# Error from server (AlreadyExists): namespaces "gatekeeper-excluded-namespace" already exists
# cleaning...
ok 10 no ignore label unless namespace is exempt test
ok 11 gatekeeper-system ignore label can be patched
ok 12 required labels dryrun test
not ok 13 container limits test
# (from function `assert_match' in file test/bats/helpers.bash, line 41,
#  in test file test/bats/test.bats, line 127)
#   `assert_match 'denied the request' "${output}"' failed
# constrainttemplate.templates.gatekeeper.sh/k8scontainerlimits unchanged
# k8scontainerlimits.constraints.gatekeeper.sh/container-must-have-limits unchanged
# checking constraint {
#     "apiVersion": "constraints.gatekeeper.sh/v1beta1",
#     "kind": "K8sContainerLimits",
#     "metadata": {
#         "annotations": {
#             "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"constraints.gatekeeper.sh/v1beta1\",\"kind\":\"K8sContainerLimits\",\"metadata\":{\"annotations\":{},\"name\":\"container-must-have-limits\"},\"spec\":{\"match\":{\"kinds\":[{\"apiGroups\":[\"\"],\"kinds\":[\"Pod\"]}],\"namespaces\":[\"gatekeeper-test-playground\"]},\"parameters\":{\"cpu\":\"200m\",\"memory\":\"1Gi\"}}}\n"
#         },
#         "creationTimestamp": "2021-02-05T18:06:27Z",
#         "generation": 1,
#         "managedFields": [
#             {
#                 "apiVersion": "constraints.gatekeeper.sh/v1beta1",
#                 "fieldsType": "FieldsV1",
#                 "fieldsV1": {
#                     "f:metadata": {
#                         "f:annotations": {
#                             ".": {},
#                             "f:kubectl.kubernetes.io/last-applied-configuration": {}
#                         }
#                     },
#                     "f:spec": {
#                         ".": {},
#                         "f:match": {
#                             ".": {},
#                             "f:kinds": {},
#                             "f:namespaces": {}
#                         },
#                         "f:parameters": {
#                             ".": {},
#                             "f:cpu": {},
#                             "f:memory": {}
#                         }
#                     }
#                 },
#                 "manager": "kubectl-client-side-apply",
#                 "operation": "Update",
#                 "time": "2021-02-05T18:06:27Z"
#             },
#             {
#                 "apiVersion": "constraints.gatekeeper.sh/v1beta1",
#                 "fieldsType": "FieldsV1",
#                 "fieldsV1": {
#                     "f:status": {
#                         ".": {},
#                         "f:auditTimestamp": {},
#                         "f:byPod": {},
#                         "f:totalViolations": {}
#                     }
#                 },
#                 "manager": "gatekeeper",
#                 "operation": "Update",
#                 "time": "2021-02-05T18:07:10Z"
#             }
#         ],
#         "name": "container-must-have-limits",
#         "resourceVersion": "2211",
#         "selfLink": "/apis/constraints.gatekeeper.sh/v1beta1/k8scontainerlimits/container-must-have-limits",
#         "uid": "d6d1b7de-e06a-4ffd-91ef-4b5eb27a812d"
#     },
#     "spec": {
#         "match": {
#             "kinds": [
#                 {
#                     "apiGroups": [
#                         ""
#                     ],
#                     "kinds": [
#                         "Pod"
#                     ]
#                 }
#             ],
#             "namespaces": [
#                 "gatekeeper-test-playground"
#             ]
#         },
#         "parameters": {
#             "cpu": "200m",
#             "memory": "1Gi"
#         }
#     },
#     "status": {
#         "auditTimestamp": "2021-02-05T18:10:06Z",
#         "byPod": [
#             {
#                 "constraintUID": "d6d1b7de-e06a-4ffd-91ef-4b5eb27a812d",
#                 "enforced": true,
#                 "id": "gatekeeper-audit-74fb6b7bf8-cws6h",
#                 "observedGeneration": 1,
#                 "operations": [
#                     "audit",
#                     "status"
#                 ]
#             },
#             {
#                 "constraintUID": "d6d1b7de-e06a-4ffd-91ef-4b5eb27a812d",
#                 "enforced": true,
#                 "id": "gatekeeper-controller-manager-688468987-8dn4x",
#                 "observedGeneration": 1,
#                 "operations": [
#                     "webhook"
#                 ]
#             },
#             {
#                 "constraintUID": "d6d1b7de-e06a-4ffd-91ef-4b5eb27a812d",
#                 "enforced": true,
#                 "id": "gatekeeper-controller-manager-688468987-k5qnq",
#                 "observedGeneration": 1,
#                 "operations": [
#                     "webhook"
#                 ]
#             }
#         ],
#         "totalViolations": 0
#     }
# }
# ready: 2, expected: 2
# expected: denied the request
# actual: The Pod "opa" is invalid: spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or `spec.tolerations` (only additions to existing tolerations)
#   core.PodSpec{
#       Volumes:        []core.Volume{{Name: "default-token-szzxm", VolumeSource: core.VolumeSource{Secret: &core.SecretVolumeSource{SecretName: "default-token-szzxm", DefaultMode: &420}}}},
#       InitContainers: nil,
#       Containers: []core.Container{
#               {
#                       ... // 6 identical fields
#                       EnvFrom: nil,
#                       Env:     nil,
#                       Resources: core.ResourceRequirements{
# -                             Limits: nil,
# +                             Limits: core.ResourceList{
# +                                     s"cpu":    {i: resource.int64Amount{value: 100, scale: -3}, s: "100m", Format: "DecimalSI"},
# +                                     s"memory": {i: resource.int64Amount{value: 31457280}, s: "30Mi", Format: "BinarySI"},
# +                             },
# -                             Requests: nil,
# +                             Requests: core.ResourceList{
# +                                     s"cpu":    {i: resource.int64Amount{value: 100, scale: -3}, s: "100m", Format: "DecimalSI"},
# +                                     s"memory": {i: resource.int64Amount{value: 31457280}, s: "30Mi", Format: "BinarySI"},
# +                             },
#                       },
#                       VolumeMounts:  []core.VolumeMount{{Name: "default-token-szzxm", ReadOnly: true, MountPath: "/var/run/secrets/kubernetes.io/serviceaccount"}},
#                       VolumeDevices: nil,
#                       ... // 11 identical fields
#               },
#       },
#       EphemeralContainers: nil,
#       RestartPolicy:       "Always",
#       ... // 25 identical fields
#   }
# cleaning...
ok 14 deployment test
ok 15 waiting for namespaces to be synced using metrics endpoint
ok 16 unique labels test
ok 17 required labels audit test
ok 18 emit events test
not ok 19 config namespace exclusion test
# (from function `wait_for_process' in file test/bats/helpers.bash, line 65,
#  in test file test/bats/test.bats, line 208)
#   `wait_for_process ${WAIT_TIME} ${SLEEP_TIME} "kubectl create configmap should-succeed -n gatekeeper-excluded-namespace"' failed
# k8srequiredlabels.constraints.gatekeeper.sh/cm-must-have-gk configured
# checking constraint {
#     "apiVersion": "constraints.gatekeeper.sh/v1beta1",
#     "kind": "K8sRequiredLabels",
#     "metadata": {
#         "annotations": {
#             "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"constraints.gatekeeper.sh/v1beta1\",\"kind\":\"K8sRequiredLabels\",\"metadata\":{\"annotations\":{},\"name\":\"cm-must-have-gk\"},\"spec\":{\"match\":{\"kinds\":[{\"apiGroups\":[\"\"],\"kinds\":[\"ConfigMap\"]}],\"namespaces\":[\"gatekeeper-test-playground\",\"gatekeeper-excluded-namespace\"]},\"parameters\":{\"labels\":[\"gatekeeper\"]}}}\n"
#         },
#         "creationTimestamp": "2021-02-05T18:06:22Z",
#         "generation": 5,
#         "managedFields": [
#             {
#                 "apiVersion": "constraints.gatekeeper.sh/v1beta1",
#                 "fieldsType": "FieldsV1",
#                 "fieldsV1": {
#                     "f:status": {
#                         ".": {},
#                         "f:auditTimestamp": {},
#                         "f:byPod": {},
#                         "f:totalViolations": {},
#                         "f:violations": {}
#                     }
#                 },
#                 "manager": "gatekeeper",
#                 "operation": "Update",
#                 "time": "2021-02-05T18:07:10Z"
#             },
#             {
#                 "apiVersion": "constraints.gatekeeper.sh/v1beta1",
#                 "fieldsType": "FieldsV1",
#                 "fieldsV1": {
#                     "f:metadata": {
#                         "f:annotations": {
#                             ".": {},
#                             "f:kubectl.kubernetes.io/last-applied-configuration": {}
#                         }
#                     },
#                     "f:spec": {
#                         ".": {},
#                         "f:match": {
#                             ".": {},
#                             "f:kinds": {},
#                             "f:namespaces": {}
#                         },
#                         "f:parameters": {
#                             ".": {},
#                             "f:labels": {}
#                         }
#                     }
#                 },
#                 "manager": "kubectl-client-side-apply",
#                 "operation": "Update",
#                 "time": "2021-02-05T18:11:00Z"
#             }
#         ],
#         "name": "cm-must-have-gk",
#         "resourceVersion": "2404",
#         "selfLink": "/apis/constraints.gatekeeper.sh/v1beta1/k8srequiredlabels/cm-must-have-gk",
#         "uid": "8db5a32b-f10a-4bb5-a285-cebdd777b751"
#     },
#     "spec": {
#         "match": {
#             "kinds": [
#                 {
#                     "apiGroups": [
#                         ""
#                     ],
#                     "kinds": [
#                         "ConfigMap"
#                     ]
#                 }
#             ],
#             "namespaces": [
#                 "gatekeeper-test-playground",
#                 "gatekeeper-excluded-namespace"
#             ]
#         },
#         "parameters": {
#             "labels": [
#                 "gatekeeper"
#             ]
#         }
#     },
#     "status": {
#         "auditTimestamp": "2021-02-05T18:10:06Z",
#         "byPod": [
#             {
#                 "constraintUID": "8db5a32b-f10a-4bb5-a285-cebdd777b751",
#                 "enforced": true,
#                 "id": "gatekeeper-audit-74fb6b7bf8-cws6h",
#                 "observedGeneration": 5,
#                 "operations": [
#                     "audit",
#                     "status"
#                 ]
#             },
#             {
#                 "constraintUID": "8db5a32b-f10a-4bb5-a285-cebdd777b751",
#                 "enforced": true,
#                 "id": "gatekeeper-controller-manager-688468987-8dn4x",
#                 "observedGeneration": 5,
#                 "operations": [
#                     "webhook"
#                 ]
#             },
#             {
#                 "constraintUID": "8db5a32b-f10a-4bb5-a285-cebdd777b751",
#                 "enforced": true,
#                 "id": "gatekeeper-controller-manager-688468987-k5qnq",
#                 "observedGeneration": 5,
#                 "operations": [
#                     "webhook"
#                 ]
#             }
#         ],
#         "totalViolations": 5,
#         "violations": [
#             {
#                 "enforcementAction": "deny",
#                 "kind": "ConfigMap",
#                 "message": "you must provide labels: {\"gatekeeper\"}",
#                 "name": "bad-cm-audit",
#                 "namespace": "gatekeeper-test-playground"
#             },
#             {
#                 "enforcementAction": "deny",
#                 "kind": "ConfigMap",
#                 "message": "you must provide labels: {\"gatekeeper\"}",
#                 "name": "bad-cm-audit-2",
#                 "namespace": "gatekeeper-test-playground"
#             },
#             {
#                 "enforcementAction": "deny",
#                 "kind": "ConfigMap",
#                 "message": "you must provide labels: {\"gatekeeper\"}",
#                 "name": "bad-cm-audit-3",
#                 "namespace": "gatekeeper-test-playground"
#             },
#             {
#                 "enforcementAction": "deny",
#                 "kind": "ConfigMap",
#                 "message": "you must provide labels: {\"gatekeeper\"}",
#                 "name": "bad-cm-audit-4",
#                 "namespace": "gatekeeper-test-playground"
#             },
#             {
#                 "enforcementAction": "deny",
#                 "kind": "ConfigMap",
#                 "message": "you must provide labels: {\"gatekeeper\"}",
#                 "name": "bad-cm-audit-5",
#                 "namespace": "gatekeeper-test-playground"
#             }
#         ]
#     }
# }
# ready: 2, expected: 2
# config.config.gatekeeper.sh/config configured
# Error from server (AlreadyExists): configmaps "should-succeed" already exists
# Error from server (AlreadyExists): configmaps "should-succeed" already exists
...
...
...
# Error from server (AlreadyExists): configmaps "should-succeed" already exists
# cleaning...
make: *** [Makefile:93: test-e2e] Error 1
```


**Environment:**

- Gatekeeper version: `v3.3.0`
- Kubernetes version: (use `kubectl version`): `v1.19.7`

--- a/test/bats/test.bats
+++ b/test/bats/test.bats
@@ -13,9 +13,11 @@ teardown() {
 }
 
 teardown_file() {
+  kubectl label ns ${GATEKEEPER_NAMESPACE} admission.gatekeeper.sh/ignore=no-self-managing --overwrite || true
   kubectl delete ns gatekeeper-test-playground gatekeeper-excluded-namespace || true
-  kubectl delete constrainttemplates k8scontainerlimits k8srequiredlabels k8suniquelabel || true
-  kubectl delete configs.config.gatekeeper.sh config -n ${GATEKEEPER_NAMESPACE} || true
+  kubectl delete "$(kubectl api-resources --api-group=constraints.gatekeeper.sh -o name | tr "\n" "," | sed -e 's/,$//')" -l gatekeeper.sh/tests=yes || true
+  kubectl delete ConstraintTemplates -l gatekeeper.sh/tests=yes || true
+  kubectl delete configs.config.gatekeeper.sh -n ${GATEKEEPER_NAMESPACE} -l gatekeeper.sh/tests=yes || true
 }
 
 @test "gatekeeper-controller-manager is running" {

--- a/test/bats/tests/constraints/all_cm_gatekeeper_label_unique.yaml
+++ b/test/bats/tests/constraints/all_cm_gatekeeper_label_unique.yaml
@@ -2,6 +2,8 @@ apiVersion: constraints.gatekeeper.sh/v1beta1
 kind: K8sUniqueLabel
 metadata:
   name: cm-gk-label-unique
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   match:
     namespaces: ["gatekeeper-test-playground"]

--- a/test/bats/tests/constraints/all_cm_must_have_gatekeeper-dryrun.yaml
+++ b/test/bats/tests/constraints/all_cm_must_have_gatekeeper-dryrun.yaml
@@ -2,6 +2,8 @@ apiVersion: constraints.gatekeeper.sh/v1beta1
 kind: K8sRequiredLabels
 metadata:
   name: cm-must-have-gk
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   enforcementAction: dryrun
   match:

--- a/test/bats/tests/constraints/all_cm_must_have_gatekeeper.yaml
+++ b/test/bats/tests/constraints/all_cm_must_have_gatekeeper.yaml
@@ -2,6 +2,8 @@ apiVersion: constraints.gatekeeper.sh/v1beta1
 kind: K8sRequiredLabels
 metadata:
   name: cm-must-have-gk
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   match:
     namespaces: ["gatekeeper-test-playground", "gatekeeper-excluded-namespace"]

--- a/test/bats/tests/constraints/all_cm_must_have_gatekeeper_audit.yaml
+++ b/test/bats/tests/constraints/all_cm_must_have_gatekeeper_audit.yaml
@@ -2,6 +2,8 @@ apiVersion: constraints.gatekeeper.sh/v1beta1
 kind: K8sRequiredLabels
 metadata:
   name: cm-must-have-gk-audit
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   match:
     namespaces: ["gatekeeper-test-playground"]

--- a/test/bats/tests/constraints/containers_must_be_limited.yaml
+++ b/test/bats/tests/constraints/containers_must_be_limited.yaml
@@ -2,6 +2,8 @@ apiVersion: constraints.gatekeeper.sh/v1beta1
 kind: K8sContainerLimits
 metadata:
   name: container-must-have-limits
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   match:
     kinds:

--- a/test/bats/tests/sync.yaml
+++ b/test/bats/tests/sync.yaml
@@ -2,6 +2,8 @@ apiVersion: config.gatekeeper.sh/v1alpha1
 kind: Config
 metadata:
   name: config
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   sync:
     syncOnly:

--- a/test/bats/tests/sync_with_exclusion.yaml
+++ b/test/bats/tests/sync_with_exclusion.yaml
@@ -2,6 +2,8 @@ apiVersion: config.gatekeeper.sh/v1alpha1
 kind: Config
 metadata:
   name: config
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   match:
     - excludedNamespaces: ["gatekeeper-excluded-namespace"]

--- a/test/bats/tests/templates/k8scontainterlimits_template.yaml
+++ b/test/bats/tests/templates/k8scontainterlimits_template.yaml
@@ -2,6 +2,8 @@ apiVersion: templates.gatekeeper.sh/v1beta1
 kind: ConstraintTemplate
 metadata:
   name: k8scontainerlimits
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   crd:
     spec:

--- a/test/bats/tests/templates/k8srequiredlabels_template.yaml
+++ b/test/bats/tests/templates/k8srequiredlabels_template.yaml
@@ -2,6 +2,8 @@ apiVersion: templates.gatekeeper.sh/v1beta1
 kind: ConstraintTemplate
 metadata:
   name: k8srequiredlabels
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   crd:
     spec:

--- a/test/bats/tests/templates/k8suniquelabel_template.yaml
+++ b/test/bats/tests/templates/k8suniquelabel_template.yaml
@@ -2,6 +2,8 @@ apiVersion: templates.gatekeeper.sh/v1beta1
 kind: ConstraintTemplate
 metadata:
   name: k8suniquelabel
+  labels:
+    gatekeeper.sh/tests: "yes"
 spec:
   crd:
     spec:

